_posts/2022-11-28-swig-for-c-and-lua.md

```diff
@@ -0,0 +1,491 @@
+    assert(type != NULL);
+    SWIG_NewPointerObj(L, my_struct, type, false);
+
+    // Call function with 1 arguments and no result
+    assert(lua_pcall(L, 1, 0, 0) == LUA_OK);
+}
+```
+
+Here, in `callProcessStruct` we first grab Lua’s `processStruct` function on top of stack. Then, we get type info for our C structure using `SWIG_TypeQuery` , followed by pushing the pointer on the stack with `SWIG_NewPointerObj`. That call creates a userdata on Lua’s stack containing our C structure. Finally, we call our Lua function.
+
+The rest of the application is just setting up the C structure with initial values, and printing the structure after call to Lua:
+
+```c
+my_struct_t my_struct;
+my_struct.isReady = true;
+my_struct.level = LEVEL_MEDIUM;
+strncpy(my_struct.message, "This is example message", sizeof(my_struct.message));
+my_struct.priority = 100;
+
+printf("[C] Calling processStruct\r\n");
+callProcessStruct(L, &my_struct);
+printf("[C] We're back from Lua\r\n");
+
+printf("[C] Printing my_struct\r\n");
+printf("\tmy_struct.level = %d\r\n", my_struct.level);
+printf("\tmy_struct.priority = %d\r\n", my_struct.priority);
+printf("\tmy_struct.message = %s\r\n", my_struct.message);
+printf("\tmy_struct.isReady = %s\r\n", my_struct.isReady ? "true" : "false"); 
+```
+
+Here’s the output:
+
+```
+[C] Calling processStruct
+[Lua] struct.level = 2.0
+[Lua] struct.priority = 100.0
+[Lua] struct.message = This is example message
+[Lua] struct.isReady = true
+[C] We're back from Lua
+[C] Printing my_struct
+        my_struct.level = 0
+        my_struct.priority = 99
+        my_struct.message = Hey from Lua
+        my_struct.isReady = false
+[C] Finished
+```
+
+It worked! Lua was able to modify our enum, integer, string and boolean. To get a peek under the hood, let’s use a well-known Lua [inspect](https://github.com/kikito/inspect.lua) library, which gives us information about objects.
+
+After importing the library with `require`, we add two lines to `processStruct`:
+
+```lua
+print("[Lua] struct: " .. inspect(struct))
+print("[Lua] struct metatable: " .. inspect(getmetatable(struct)))
+```
+
+And get the following print (truncated):
+
+```
+[Lua] struct: <userdata 1>
+[Lua] struct metatable: <1>{
+...
+  [".get"] = {
+    isReady = <function 2>,
+    level = <function 3>,
+    message = <function 4>,
+    priority = <function 5>
+  },
+	[".set"] = {
+    isReady = <function 6>,
+    level = <function 7>,
+    message = <function 8>,
+    priority = <function 9>
+  },
+...
+```
+
+You can see from Lua point of view, the object is Lua’s [userdata](https://www.lua.org/pil/28.1.html), which means it’s raw memory. Normally userdata is managed by Lua meaning it will be garbage collected. However, SWIG gives us flexibility here with its `SWIG_NewPointerObj` API. The last argument to that function specified the owner of the object, and since we decided to own the object the garbage collector won’t affect it.
```

Comment:
```suggestion
You can see from Lua's point of view, the object is Lua’s [userdata](https://www.lua.org/pil/28.1.html), which means it’s raw memory. Normally userdata is managed by Lua meaning it will be garbage collected. However, SWIG gives us flexibility here with its `SWIG_NewPointerObj` API. The last argument to that function specified the owner of the object, and since we decided to own the object the garbage collector won’t affect it.
```